## Part 1: The Doomsday Algorithm

The Doomsday algorithm, devised by mathematician J. H. Conway, computes the day of the week any given date fell on. The algorithm is designed to be simple enough to memorize and use for mental calculation.

__Example.__ With the algorithm, we can compute that July 4, 1776 (the day the United States declared independence from Great Britain) was a Thursday.

The algorithm is based on the fact that for any year, several dates always fall on the same day of the week, called the <em style="color:#F00">doomsday</em> for the year. These dates include 4/4, 6/6, 8/8, 10/10, and 12/12.

__Example.__ The doomsday for 2016 is Monday, so in 2016 the dates above all fell on Mondays. The doomsday for 2017 is Tuesday, so in 2017 the dates above will all fall on Tuesdays.

The doomsday algorithm has three major steps:

1. Compute the anchor day for the target century.
2. Compute the doomsday for the target year based on the anchor day.
3. Determine the day of week for the target date by counting the number of days to the nearest doomsday.

Each step is explained in detail below.

### The Anchor Day

The doomsday for the first year in a century is called the <em style="color:#F00">anchor day</em> for that century. The anchor day is needed to compute the doomsday for any other year in that century. The anchor day for a century $c$ can be computed with the formula:
$$
a = \bigl( 5 (c \bmod 4) + 2 \bigr) \bmod 7
$$
The result $a$ corresponds to a day of the week, starting with $0$ for Sunday and ending with $6$ for Saturday.

__Note.__ The modulo operation $(x \bmod y)$ finds the remainder after dividing $x$ by $y$. For instance, $12 \bmod 3 = 0$ since the remainder after dividing $12$ by $3$ is $0$. Similarly, $11 \bmod 7 = 4$, since the remainder after dividing $11$ by $7$ is $4$.

__Example.__ Suppose the target year is 1954, so the century is $c = 19$. Plugging this into the formula gives
$$a = \bigl( 5 (19 \bmod 4) + 2 \bigr) \bmod 7 = \bigl( 5(3) + 2 \bigr) \bmod 7 = 3.$$
In other words, the anchor day for 1900-1999 is Wednesday, which is also the doomsday for 1900.

__Exercise 1.1.__ Write a function that accepts a year as input and computes the anchor day for that year's century. The modulo operator `%` and functions in the `math` module may be useful. Document your function with a docstring and test your function for a few different years.  Do this in a new cell below this one.

In [1]:
def year_to_anchor(year):
    c = year//100
    a = (5*(c%4)+2)%7
    return a



### The Doomsday

Once the anchor day is known, let $y$ be the last two digits of the target year. Then the doomsday for the target year can be computed with the formula:
$$d = \left(y + \left\lfloor\frac{y}{4}\right\rfloor + a\right) \bmod 7$$
The result $d$ corresponds to a day of the week.

__Note.__ The floor operation $\lfloor x \rfloor$ rounds $x$ down to the nearest integer. For instance, $\lfloor 3.1 \rfloor = 3$ and $\lfloor 3.8 \rfloor = 3$.

__Example.__ Again suppose the target year is 1954. Then the anchor day is $a = 3$, and $y = 54$, so the formula gives
$$
d = \left(54 + \left\lfloor\frac{54}{4}\right\rfloor + 3\right) \bmod 7 = (54 + 13 + 3) \bmod 7 = 0.
$$
Thus the doomsday for 1954 is Sunday.

__Exercise 1.2.__ Write a function that accepts a year as input and computes the doomsday for that year. Your function may need to call the function you wrote in exercise 1.1. Make sure to document and test your function.

In [4]:
def year_to_doomsday(year):
    a = year_to_anchor(year)
    y = year%100
    d = (y + (y//4) + a)%7
    return d

In [5]:
year_to_doomsday(2001)

3

### The Day of Week

The final step in the Doomsday algorithm is to count the number of days between the target date and a nearby doomsday, modulo 7. This gives the day of the week.

Every month has at least one doomsday:
* (regular years) 1/10, 2/28
* (leap years) 1/11, 2/29
* 3/21, 4/4, 5/9, 6/6, 7/11, 8/8, 9/5, 10/10, 11/7, 12/12

__Example.__ Suppose we want to find the day of the week for 7/21/1954. The doomsday for 1954 is Sunday, and a nearby doomsday is 7/11. There are 10 days in July between 7/11 and 7/21. Since $10 \bmod 7 = 3$, the date 7/21/1954 falls 3 days after a Sunday, on a Wednesday.

__Exercise 1.3.__ Write a function to determine the day of the week for a given day, month, and year. Be careful of leap years! Your function should return a string such as "Thursday" rather than a number. As usual, document and test your code.

In [6]:
def date_to_weekday(month, day, year):
    d = year_to_doomsday(year)
    
    #leap years
    #assuming tanner is correct (see slack #questions)
    if (month == 1 or month == 2) and ((year%4 == 0 and not year%100 == 0) or year%400 == 0):
        day = day-1

    #now for the month specific doomsdates
    if month == 1:
        x = (day-10)%7
    if month == 2:
        x = (day-28)%7
    if month == 3:
        x = (day-21)%7
    if month == 4:
        x = (day-4)%7
    if month == 5:
        x = (day-9)%7
    if month == 6:
        x = (day-6)%7
    if month == 7:
        x = (day-11)%7
    if month == 8:
        x = (day-8)%7
    if month == 9:
        x = (day-5)%7
    if month == 10:
        x = (day-10)%7
    if month == 11:
        x = (day-7)%7
    if month == 12:
        x = (day-12)%7
    if month < 1 or month > 12:
        print("invalid month")
        return None
    
    #for determining day code
    daycode = (x + d)%7
    
    #for determining day
    if daycode == 0:
        wd = "Sunday"
    if daycode == 1:
        wd = "Monday"
    if daycode == 2:
        wd = "Tuesday"
    if daycode == 3:
        wd = "Wednesday"
    if daycode == 4:
        wd = "Thursday"
    if daycode == 5:
        wd = "Friday"
    if daycode == 6:
        wd = "Saturday"
    
    return wd


__Exercise 1.4.__ How many times did Friday the 13th occur in the years 1900-1999? Does this number seem to be similar to other centuries?

In [7]:
def count_F13(a,b):
    tally=0
    for year in range(a,b):
        for month in range(1,13):
            if date_to_weekday(month, 13, year) == "Friday":
                tally = tally + 1
    return tally

In [8]:
print(count_F13(1900,2000))
print(count_F13(1800,1900),count_F13(1700,1800),count_F13(1600,1700),count_F13(1500,1600),
      count_F13(1400,1500),count_F13(1300,1400),count_F13(1200,1300),count_F13(1100,1200))

172
172 172 172 172 172 172 172 172


##### There are 172 Friday the 13th's in the years 1900-1999. This number seems similar to other centuries

__Exercise 1.5.__ How many times did Friday the 13th occur between the year 2000 and today?

In [9]:
print(count_F13(2000,2018))

31


##### There have been 31 occurances of Friday the 13th since the start of this century.

## Part 2: 1978 Birthdays

__Exercise 2.1.__ The file `birthdays.txt` contains the number of births in the United States for each day in 1978. Inspect the file to determine the format. Note that columns are separated by the tab character, which can be entered in Python as `\t`. Write a function that uses iterators and list comprehensions with the string methods `split()` and `strip()` to  convert each line of data to the list format

```Python
[month, day, year, count]
```
The elements of this list should be integers, not strings. The function `read_birthdays` provided below will help you load the file.

In [10]:
def read_birthdays(file_path):
    """Read the contents of the birthdays file into a string.
    
    Arguments:
        file_path (string): The path to the birthdays file.
        
    Returns:
        string: The contents of the birthdays file.
    """
    with open(file_path) as file:
        return file.read()
    

In [11]:
import string

In [12]:
def lines_to_list():
    bdays=read_birthdays("birthdays.txt")
    bdays.strip()
    blist = bdays.split("\n")
    #seventh line is the first birthday, line 370 is last birthday
    blist = blist[6:371]
    biter = iter(blist)

    
    #now do idividual lists
    for i in range(365):
        bday = next(biter)     #get the next bithday on the list
        date, count = bday.split("\t")     #split at the tab to make date and count
        mon,day,yr = date.split("/")     #split at the / to separate date into month, day, year
        
        mon,day,yr,count = int(mon),int(day),int(yr),int(count)     #cast all strings as ints
        blist[i] = [mon,day,yr,count]     #put the newly formated birthday back into the list
        
    return(blist)

__Exercise 2.2.__ Which month had the most births in 1978? Which day of the week had the most births? Which day of the week had the fewest? What conclusions can you draw? You may find the `Counter` class in the `collections` module useful.

In [13]:
from collections import Counter

In [14]:
x = lines_to_list() 
month_counts = sum((Counter({r[0]:r[3]}) for r in x),Counter())
month_counts

Counter({1: 270695,
         2: 249875,
         3: 276584,
         4: 254577,
         5: 270812,
         6: 270756,
         7: 294701,
         8: 302795,
         9: 293891,
         10: 288955,
         11: 274671,
         12: 284927})

In [15]:
x = lines_to_list()
weekdays=[]
i=0
for line in x:
    weekday = date_to_weekday(line[0],line[1],line[2])
    number = line[3]
    weekdays.append([weekday,number])
    i+=1

sum((Counter({r[0]:r[1]}) for r in weekdays),Counter())

Counter({'Friday': 432085,
         'Monday': 504858,
         'Saturday': 421400,
         'Sunday': 487309,
         'Thursday': 500541,
         'Tuesday': 493897,
         'Wednesday': 493149})

##### August has the most births of any month. Monday has the most births of any day of the week. 
##### Saturday has the fewest births of any day of the week.

__Exercise 2.3.__ What would be an effective way to present the information in exercise 2.2? You don't need to write any code for this exercise, just discuss what you would do.

##### A contigency table would be my first instict. Graphically, I think maybe a histogram or a pie chart. 